src/workflow_provider/default_vm_workflow_provider.cpp

```diff
@@ -0,0 +1,263 @@
+#include <multipass/exceptions/invalid_memory_size_exception.h>
+#include <multipass/exceptions/workflow_exceptions.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/query.h>
+#include <multipass/url_downloader.h>
+#include <multipass/utils.h>
+
+#include <yaml-cpp/yaml.h>
+
+#include <QFile>
+#include <QFileInfo>
+
+#include <Poco/StreamCopier.h>
+#include <Poco/Zip/ZipArchive.h>
+#include <Poco/Zip/ZipStream.h>
+
+#include <fstream>
+#include <sstream>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+const QString github_workflows_archive_name{"multipass-workflows.zip"};
+const QString workflow_dir_version{"v1"};
+constexpr auto category = "workflow provider";
+
+auto workflows_map_for(const std::string& archive_file_path)
+{
+    std::map<std::string, std::string> workflows_map;
+    std::ifstream zip_stream{archive_file_path, std::ios::binary};
+    Poco::Zip::ZipArchive zip_archive{zip_stream};
+
+    for (auto it = zip_archive.headerBegin(); it != zip_archive.headerEnd(); ++it)
+    {
+        if (it->second.isFile())
+        {
+            auto file_name = it->second.getFileName();
+            QFileInfo file_info{QString::fromStdString(file_name)};
+
+            if (file_info.path().contains(workflow_dir_version) && file_info.suffix() == "yaml")
+            {
+                Poco::Zip::ZipInputStream zip_input_stream{zip_stream, it->second};
+                std::ostringstream out(std::ios::binary);
+                Poco::StreamCopier::copyStream(zip_input_stream, out);
+                workflows_map[file_info.baseName().toStdString()] = out.str();
+            }
+        }
+    }
+
+    return workflows_map;
+}
+} // namespace
+
+mp::DefaultVMWorkflowProvider::DefaultVMWorkflowProvider(const QUrl& workflows_url, URLDownloader* downloader,
+                                                         const QDir& archive_dir,
+                                                         const std::chrono::milliseconds& workflows_ttl)
+    : workflows_url{workflows_url},
+      url_downloader{downloader},
+      archive_file_path{archive_dir.filePath(github_workflows_archive_name)},
+      workflows_ttl{workflows_ttl}
+{
+    update_workflows();
+}
+
+mp::DefaultVMWorkflowProvider::DefaultVMWorkflowProvider(URLDownloader* downloader, const QDir& archive_dir,
+                                                         const std::chrono::milliseconds& workflows_ttl)
+    : DefaultVMWorkflowProvider(default_workflow_url, downloader, archive_dir, workflows_ttl)
+{
+}
+
+mp::Query mp::DefaultVMWorkflowProvider::fetch_workflow_for(const std::string& workflow_name,
+                                                            VirtualMachineDescription& vm_desc)
+{
+    update_workflows();
+
+    Query query{"", "", false, "", Query::Type::Alias};
+    auto& config = workflow_map.at(workflow_name);
+    auto workflow_config = YAML::Load(config);
```

Comment:
Feels weird to be parsing it JIT, there's a lot of `YAML::Load(config)` below. I suppose it depends what we want to optimize for - memory or CPU usage…

What we should do, though, is handle any errors in the YAML. Everywhere below you assume the YAML is correct with our schema. But what if it's not? Some more tests for that would be in order, too.

Reply:
Sure, it's certainly possible that the map stores the YAML node instead.

And yes, we should catch YAML exceptions when converting to a type, log it and invalidate the particular workflow due to the error.

Reply:
I've (hopefully) addressed the handling of errors now.

Regarding the current JIT approach, I'm wondering if it should store the loaded `YAML::Node` in the map or actually parse the YAML and store the associated data structures (like `VirtualMachineDescription`, etc.) in the map?  WDYT?

Reply:
Either way works… No preference, really.

Reply:
Ok, pushed a commit to just store the `YAML::Node` in the map.  We can optimize further later if need be.